src/LeastSquaresFit.cpp

```diff
@@ -4,17 +4,75 @@
 #include <iostream>
 #include <utility> //std::pair
 #include <numeric> // std::inner_product
+#include <chrono>
+
+using vec_col_float = std::vector<ColumnData<float_data_type>>;
+using vec_col_int = std::vector<ColumnData<int_data_type>>;
+
+template <typename X_type, typename Y_type>
+void runProfile(const index_type num_segments, const CSVParser &csv,
+		const std::string &X_header, const std::string &Y_header,
+		const index_type header_row_index, 	const index_type first_data_row_index) {
+
+	using vec_cols_tX = std::vector<ColumnData<X_type>>;
+	using vec_cols_tY = std::vector<ColumnData<Y_type>>;
+	using time_nano_t = std::chrono::nanoseconds;
+
+	// If num_segments != 1, run the analysis with a single segment (non-progressive)
+	// to obtain the true y-intercept and slope.  Use these values to determine
+	// the error during progressive calculations
+	float_data_type a_actual, b_actual;
+	if(num_segments != 1) {
```

Comment:
I'm not sure I can see a way to use an optional here since there is a required argument (num_segments) that is provided to this function.  When I looked at this function again, I figured perhaps it's fine to just leave that if statement out altogether.  If I'm profiling a single-segment calculation, it will just do it twice.  Not a big deal, and it doesn't affect the profiling.

Any opposition to me just leaving this if statement out altogether?  Leaving it out also fixes my currently wrong error value when I profile a single segment calculation.

Reply:
I should have given more details, sorry. The function signature would change to:
```
template <typename X_type, typename Y_type>
void runProfile(const std::optional< index_type > num_segments, const CSVParser &csv,
		const std::string &X_header, const std::string &Y_header,
		const index_type header_row_index, const index_type first_data_row_index);
```
And then this if statement would change to:
```
if(num_segments.has_value()) {
```
The later use of the value would simply change to:
```
	auto Xs = csv.makeSegments<X_type>(X_header, 
		header_row_index, first_data_row_index, *num_segments);
```

And the call site for progressive computation would remain unchanged but the call site (`main()`) for non-progressive computation you could change `num_segs` to an optional (rather than required) input parameter and initialise its value to `std::nullopt`
```

You get a few advantages to this (other than just showing off your knowledge of c++17! :smile: ):

  * It frees you to use an unsigned type for your indexes, because you no longer need to reserve a special value for your sentinel `-1`: that means you can handle datasets twice as large before needing to increase the storage.
 * It is clearer both to the user (who can choose not specify `num_segs` on the command line) and the programmer (who no longer needs multi-line comments to understand the meaning of the condition) to use the `std::optional<>` over a `-1` sentinel value.
 * You don't double the waiting time for the non-progressive computation.

But as with any feedback, _your kilometrage may vary_. :smile: